Refactor `UpgradeATStep` tests to @testing-library/react

brookewp opened the pull request:
#### Proposed Changes

* This PR refactors the `UpgradeATStep` component to use `@testing-library/react` instead of `enzyme`.

#### Testing Instructions

<!--
Add as many details as possible to help others reproduce the issue and test the fix.
"Before / After" screenshots can also be very helpful when the change is visual.
-->

* Verify tests still pass: `yarn test-client client/components/marketing-survey/cancel-purchase-form/step-components/test/upgrade-at-step.jsx`

<!--
Link a related issue to this PR. If the PR does not immediately resolve the issue,
for example, it requires a separate deployment to production, avoid
using the "fixes" keyword and instead attach the [Status] Fix Inbound label to
the linked issue.
-->

Related to #63409

brookewp committed:
migrate tests from enyzme to testing-library

brookewp commented on client/components/marketing-survey/cancel-purchase-form/step-components/test/upgrade-at-step.jsx:
```diff
@@ -1,67 +1,57 @@
+		const btn = screen.queryByRole( 'link' );
 
-		wrapper.find( Button ).simulate( 'click' );
+		btn.addEventListener( 'click', ( event ) => event.preventDefault(), false );
```
I ran into `Error: Not implemented: navigation (except hash changes)` when trying to simulate a click on the anchor. I implemented the above solution found here: https://github.com/jsdom/jsdom/issues/2112#issuecomment-926601210

tyxla replied:
Thanks for this, it's a really clean way to disable navigation. Another would be to mock `window.location`, but yours seems cleaner IMHO 👍

tyxla commented on client/components/marketing-survey/cancel-purchase-form/step-components/test/upgrade-at-step.jsx:
```diff
@@ -1,67 +1,57 @@
+		render( <UpgradeATStep { ...props } recordTracksEvent={ noop } translate={ noop } /> );
 
-		expect( wrapper.find( Button ).props().href ).toEqual(
+		expect( screen.queryByRole( 'link' ) ).toHaveAttribute(
```
I wonder if we're entering into the implementation detail territory here (caused by the original test, of course). The thing is, `Button` will decide whether to render a link or a button, and we shouldn't be testing its details - this sounds more like job for `Button`'s own tests. Is it a better idea to mock `Button` and just ensure that it's receiving the props we expect it to receive? Or should we just accept this as-is because it brings value as-is, and take the fact that it will be more like an integration test rather than a unit test?

brookewp replied:
That's a good point! If we mock the button, is there a way we can do so without affecting the former test that checks for the translated content? Or a way to include the translated content into the mock? 🤔

tyxla replied:
What we've done before is to mock `useTranslate()` or `translate()` to return or resemble a function that returns its first argument (just like lodash's `identity` works - https://lodash.com/docs/4.17.15#identity). Would that work in your case?

tyxla reviewed (approved):
Looking great! One thing we could potentially improve would be how we test details that bleed into other components and are not a responsibility of this component to test. We could make a compromise and treat tests as more integration or closer to an e2e test, but we could also limit the responsibility and test only what our component is responsible for, nothing more. Up to you - nothing of that is blocking anyway! 🚀

brookewp committed:
Update tests based on PR feedback

brookewp commented:
Thanks @tyxla for your original review! I've added a mock for the button - let me know if that's what you were thinking or if I misunderstood. 

I also incorporated your feedback from [my other similar PR](https://github.com/Automattic/wp-calypso/pull/65140), in regards to the [translate mock](https://github.com/Automattic/wp-calypso/pull/65140/files#r913615220) and [userEvent.setup()](https://github.com/Automattic/wp-calypso/pull/65140/files#r913616548).

tyxla reviewed (approved):
LGTM, thanks 🚀
